src/pip/_internal/index/collector.py

```diff
@@ -42,6 +43,35 @@
 logger = logging.getLogger(__name__)
 
 
+# TODO: This is a stand-in for @functools.lru_cache() until python 2 support is
+# dropped. mypy does not appear to understand `except ImportError`, so this
+# polyfill is used everywhere.
+def _lru_cache(
+        *args,                  # type: Any
+        **kwargs                # type: Any
+):
+    # type: (...) -> Any
+    cache = {}                  # type: Dict[Any, Any]
+
+    def wrapper(fn):
+        # type: (Any) -> Any
+
+        def wrapped(
+                *args,          # type: Any
+                **kwargs        # type: Any
+        ):
+            # type: (...) -> Any
+            cache_key = tuple(args) + tuple(kwargs.items())
+            value = cache.get(cache_key, None)
+            if value is not None:
+                return value
+            value = fn(*args, **kwargs)
+            cache[cache_key] = value
+            return value
+        return wrapped
+    return wrapper
```

Comment:
I wonder whether we should simply make the cache a no-op on Python 2. Something like:

```python
F = TypeVar('F', Callable)

def stub_lru_cache(maxsize=None):
    # type: (Optional[int]) -> Callable[[F], F]
    def _wrapper(f):
        # type: (F) -> F
        return f
    return _wrapper
try:
    from functools import lru_cache as real_lru_cache
except ImportError:
    _lru_cache = stub_lru_cache
else:
    _lru_cache = real_lru_cache
```

There are probably a couple of typing hoops we need to jump through, but the point is not avoid vendoring a duplicated `lru_cache` altogether.

Reply:
A couple of references for convinience, if we are going down this route:

* python/mypy#1297 How to deal with conditional imports
* python/mypy#5107 How to type-check `lru_cache`

Reply:
I have slightly modified your example to make mypy and flake8 accept it: https://github.com/pypa/pip/blob/83153538e61e056b9b99f5334684e7001081a90a/src/pip/_internal/index/collector.py#L49-L57

Reply:
I pushed a patch against your repo to replace the `type: ignore` comments with `getattr()` and a callback protocol that IMO is easier to follow. You can merge that PR to include it in this PR.

Reply:
Thank you for diving into this and figuring out the Protocol usage!! I have merged the PR!